Article about solving contento.me (#103)
qdrant/landing_page: qdrant-landing/content/articles/solving-contexto.md

---
title: Solving Contexto.me with Vector Search
short_description: Solving Contexto.me with Vector Search and its practical implications
description: How to solve Contexto.me with Vector Search and why it is more important than it seems
social_preview_image: /articles_data/solving-contexto/preview/social_preview.jpg
preview_dir: /articles_data/solving-contexto/preview
small_preview_image: /articles_data/solving-contexto/icon.svg
weight: 8
author: Andrei Vasnetsov
author_link: https://blog.vasnetsov.com/

> date: 2022-06-28T08:57:07.604Z

Reviewer: old date

Author: too long for title

# aliases: [ /articles/solving-contexto/ ]
---

<!---

Plan:

- What is Contexto.me, how it works

> - Naive aproaches and why they don't work

Reviewer: ```suggestion
- Naive approaches and why they don't work
```

-->

[Contexto.me](https://contexto.me/) is a linguistic game that takes the popular word game [Wordle](https://www.nytimes.com/games/wordle/index.html) to the next level.
In this game, players must guess a secret word by submitting guesses and receiving feedback on the similarity of their guess to the secret word.

The game claims that it "uses an artificial intelligence algorithm to sort words by their similarity to the secret word".
When a player submits a guess, they receive feedback on its position in the sorted list of words.
Players have an unlimited number of guesses, but the game rewards those who can solve it with fewer attempts.

Try to solve it yourself and then come back to see how we tough the machine to solve it!

## Naive approaches

<!---

There are naive approaches we tried:

- The game is obviously using some kind of Word2Vec model to sort words by similarity. 
- Explain what Word2Vec is.

- We can start with random word and just look into the list of words similar to it. If we see a word that is close to the secret word, we use it as a reference. Then we look into the list of words similar to the reference word and so on. If we see the secret word, we stop.

> - This approach works, but it is very slow. It tends to stuck in clusters of words that are similar to each other, which forces us to retrive a lot of words from the model.

Reviewer: ```suggestion
- This approach works, but it is very slow. It tends to stuck in clusters of words that are similar to each other, which forces us to retrieve a lot of words from the model.
```

- We also can't use tricks from linear algebra to evaluate exact vector based on distances to given points.
    - First, because we don't know exact word2vec model which was used to sort words.
    - Second, because we don't know the exact distance to the secret word.
    - We can only compare distances between words.

> The solution should not only account for the most sililar word we found so far, but also consider the distance to the words it found to be unsimilar.

Reviewer: ```suggestion
The solution should not only account for the most similar word we found so far, but also consider the distance to the words it found to be dissimilar.
```

{{< figure src=/articles_data/solving-contexto/cbow-word2vec.webp caption="Word2Vec training architecture">}}

Word2vec is one of the first methods used to represent objects in vector space. 
Currently there are a lot of more sophisticated methods, that can capture meaning of the whole texts, not just a single word.
But for our purposes word2vec will work just fine.

So, here's the naive approach you've probably already thought of:

We can start with a random word and look into the list of similar words using some Word2Vec model.
If we see a word closer to the secret word, we use it as a reference and repeat the process.

Although this approach works if we are initially close enough to the secret word, it is generally quite slow and inefficient.
It tends to get stuck in clusters of words that are similar to each other, forcing us to retrieve many words from the model.

Additionally, using linear algebra techniques to evaluate the exact vector based on distances to given points does not look feasible in this scenario.

> This is because the exact word2vec model used to sort the words is unknown, as is the exact distance to the secret word.

Reviewer 2: It was unclear to me what "exact word2vec model" means until I read the prompt put into ChatGPT. Maybe we could replace it with the "original word2vec model"?

The only option is to compare distances between words.

## One working approach

Based on the previous section, we can conclude, that using only the most similar word found so far is not enough to generate efficient guesses. 
The more efficient solution must also consider the words deemed dissimilar.

Let's consider the simplest case: we have guessed 2 words `house` and `blue` and received feedback on their similarity to the secret word.

One of the words is closer to the secret word than the other, so we can make some assumptions about the secret word.

> We understand that the secret word is more likely to be similar to `house` than `blue`, but we only have the informaion about its relative similarity to these two words.

Reviewer: ```suggestion
We understand that the secret word is more likely to be similar to `house` than `blue`, but we only have the information about its relative similarity to these two words.
```

Let's assign a score to each word in the vocabulary based on this observation:

{{< figure src=/articles_data/solving-contexto/scoring-1.png caption="Scoring words based on 2 guesses">}}

We assign +1 score to those words that are closer to `house` than `blue` and -1 score to those words that are closer to `blue` than `house`.

Now, we can use this score to rank the words in the vocabulary and use word with the highest score as our next guess.

Let's see how scores change after we make a third guess:

{{< figure src=/articles_data/solving-contexto/scoring-2.png caption="Ranking words based on next 2 guesses">}}

We can generalize this approach to any number of guesses.

> The simpliest way to do this is to sample pairs of guesses and update the score iteratively.

Reviewer: ```suggestion
The simplest way to do this is to sample pairs of guesses and update the score iteratively.
```

> That's it! We can use this approach to suggest words one by one and extend guess list accordingly.

Reviewer 2: I actually expected the approach would be to use the recommendation API, or even just search with all the embeddings stored in Qdrant. That might be an interesting case to promote, but there is no reference to Qdrant, while there could be.

Author: there is reference in the last section

Benefits of this approach:

> - It is stochastic. if there are inconsistencies in the input data, the algorithm can tolerate them.

Reviewer: Either use colon/comma or capitalize if
```suggestion
- It is stochastic. If there are inconsistencies in the input data, the algorithm can tolerate them.
```

- The algorithm does not require using exactly the same model as used in the game. It can work with any distance metric and any dimensionality of the vector space.
- The algorithm is invariant to the order of the input data.
- Algorithm only relies on the relative similarity of the words and can be easily adapted to other types of input.

> We even made a simple sctipt you that you can run youself, check it out on [GitHub](https://github.com/qdrant/contexto).

Reviewer: ```suggestion
We even made a simple script you that you can run yourself, check it out on [GitHub](https://github.com/qdrant/contexto).
```
